openwisp_controller/connection/api/views.py

```
@@ -131,15 +133,38 @@ def get_serializer_context(self):
         context["device_id"] = self.kwargs["device_id"]
         return context
 
+    def initial(self, *args, **kwargs):
+        super().initial(*args, **kwargs)
+        self.assert_parent_exists()
+
+    def assert_parent_exists(self):
+        try:
+            assert self.get_parent_queryset().exists()
+        except (AssertionError, ValidationError):
```

Ruff (0.15.0)

[error] Undefined name `ValidationError`

(F821)